### PR TITLE
gh-99064: Added warnings for file objects with non-UTF8 content

### PR DESCRIPTION
**Pull request by rruuaanng**
https://github.com/python/cpython/issues/99064 mentioned that if source contains non-ASCII characters, the parse method will return incorrect results. However, there is no warning for this process.

* Issue: gh-99064

### CODE REVIEW

**Comment by erlend-aasland** [comment]
Please see the [Making good PRs](https://devguide.python.org/getting-started/pull-request-lifecycle/#making-good-prs) section of the devguide. I suggest taking a close look at it before opening new PRs; incomplete PRs only result in wasted reviewer and CI churn.

This PR lacks documentation and tests.

**Comment by rruuaanng** [comment]
I have made changes, Please reviewer! @erlend-aasland

File: Lib/xml/etree/ElementTree.py [hunk 1]
```diff
@@ -1200,6 +1200,15 @@ def parse(source, parser=None):
     Return an ElementTree instance.
 
     """
+    import warnings
+
+    if (hasattr(source, 'encoding') and
+        source.encoding == 'ISO-8859-1' and
+        source.mode == 'r'):
+        warnings.warn(
```
Comment (vstinner) [inline Lib/xml/etree/ElementTree.py]: Please move the "import warnings" here.

File: Misc/NEWS.d/next/Library/2024-09-10-11-19-05.gh-issue-99064.iHNbcr.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Added a warning in the :func:`xml.etree.ElementTree.parse` for
+``'ISO-8859-1'`` encoding files opened in ``r`` mode. Patch by RUANG.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2024-09-10-11-19-05.gh-issue-99064.iHNbcr.rst]:
```suggestion
Added a warning in the :func:`xml.etree.ElementTree.parse` for
files opened in ``r`` mode with ``'ISO-8859-1'`` encoding. Patch by RUANG.
```

File: Lib/xml/etree/ElementTree.py (hunk 1 above)
Comment (rruuaanng) [reply Lib/xml/etree/ElementTree.py]: It seems more perfect this way. I've made changes :)

Comment (erlend-aasland) [reply Lib/xml/etree/ElementTree.py]: Moreover, why not just `if getattr(source, 'encoding', None) == 'ISO-8859-1' and source.mode == 'r'`?

**Comment by rruuaanng** [comment]
Thank for @vstinner, @erlend-aasland reviewer. Thank you for your hard work! Please review again :)

File: Lib/xml/etree/ElementTree.py [hunk 3]
```diff
@@ -1200,6 +1200,13 @@ def parse(source, parser=None):
     Return an ElementTree instance.
 
     """
+    if (getattr(source, 'encoding', None) == 'ISO-8859-1' and
+        source.mode == 'r'):
+        import warnings
+        warnings.warn(
+            "For file objects containing XML data"
```
Comment (erlend-aasland) [inline Lib/xml/etree/ElementTree.py]:
```suggestion
            "For file objects containing XML data "
```

File: Lib/xml/etree/ElementTree.py [hunk 4]
```diff
@@ -1200,6 +1200,13 @@ def parse(source, parser=None):
...
+            "with non-ASCII and non-UTF-8 encoding (e.g. ISO 8859-1), "
```
Comment (erlend-aasland) [inline Lib/xml/etree/ElementTree.py]: We try to avoid latin abbreviations. Please spell it out in plain english.

Comment (rruuaanng) [reply Lib/xml/etree/ElementTree.py]: It seems that e.g. is more concise than for example (although, I agree with your suggestion).

File: Lib/xml/etree/ElementTree.py [hunk 5]
```diff
@@ -1200,6 +1200,13 @@ def parse(source, parser=None):
     Return an ElementTree instance.
 
     """
+    if (getattr(source, 'encoding', None) == 'ISO-8859-1' and
+        source.mode == 'r'):
```
Comment (erlend-aasland) [inline Lib/xml/etree/ElementTree.py]: IMO, this is a case where 80 chars is ok.

```suggestion
    if getattr(source, 'encoding', None) == 'ISO-8859-1' and source.mode == 'r':
```

**Comment by rruuaanng** [comment]
They may not have any other problems. Thank for reviewer.

File: Doc/library/xml.etree.elementtree.rst [hunk 6]
```diff
@@ -665,6 +665,11 @@ Functions
    not given, the standard :class:`XMLParser` parser is used.  Returns an
    :class:`ElementTree` instance.
 
+   .. warning::
+
+      When the *source* encoding is ``ISO-8859-1`` and the mode is ``r``, An encoding
+      error warning will be thrown.
```
Comment (vstinner) [inline Doc/library/xml.etree.elementtree.rst]:
```suggestion
   .. note::

      When the *source* encoding is ``ISO-8859-1`` and the mode is ``r``,
      a warning is emited.
```

Comment (rruuaanng) [reply Doc/library/xml.etree.elementtree.rst]: Oh, I was wondering if using warnings would be better. But now I know.

Comment (vstinner) [reply Doc/library/xml.etree.elementtree.rst]: I don't think that ``.. warnings::`` is justified here, it's not a major issue.

File: Lib/xml/etree/ElementTree.py [hunk 7]
```diff
@@ -1200,6 +1200,12 @@ def parse(source, parser=None):
     Return an ElementTree instance.
 
     """
+    if getattr(source, 'encoding', None) == 'ISO-8859-1' and source.mode == 'r':
+        import warnings
+        warnings.warn(
+            "For file objects containing XML data "
+            "with non-ASCII and non-UTF-8 encoding (for example. ISO 8859-1), "
+            "the file must have been opened in binary mode.", category=RuntimeWarning)
```
Comment (vstinner) [inline Lib/xml/etree/ElementTree.py]: Honestly, I don't understand this warning message :-( I am not sure how to rephase it.

The warning is emitted even if XML is encoded to ASCII. "Non-UTF-8" is unclear to me. I'm not sure why ISO 8859-1 is only given as an example, whereas the warning is only emitted if the encoding is ISO-8859-1.

Should we also emit the warning for ISO-8859-15? What about ShiftJIS? Or Big5?

Let me try to propose a better message.

"File using {source.encoding} encoding should open XML in binary mode."

And consider emitting the warning if the encoding is not ASCII nor UTF-8.

I'm not a XML expert, so I'm not sure :-(

Comment (rruuaanng) [reply Lib/xml/etree/ElementTree.py]: In fact, `non-UTF8` and `non-ASCII` refer to something similar to **ISO-8859-1**, But in fact the character set it contains is not similar to `ASCII`, and it isn't considered `UTF8`. I agree with your point of view, it should be shortened, I think it should be changed to `ISO 8859-1 encoding should be read using 'rb' mode`.

**Comment by rruuaanng** [comment]
Maybe I should introduce **ISO-8859-1**. It adds `96` letters and symbols to the vacant range of `0xA0-0xFF` for use by Latin alphabet languages ​​that use additional symbols. But in fact, the newly added part does not belong to `UTF8`, and since it adds 96 new contents, it is not actually `ASCII`. Since it is an **8-bit** character set, it must be read byte by byte.
for more to see https://en.wikipedia.org/wiki/ISO/IEC_8859-1.

File: Lib/xml/etree/ElementTree.py (hunk 7 above)
Comment (erlend-aasland) [reply Lib/xml/etree/ElementTree.py]: > Should we also emit the warning for ISO-8859-15? What about ShiftJIS? Or Big5?

IMO, that should be identified (on the issue) first. It seems weird to only add a warning for one particular non-UTF8 encoding. I would guess that the problem is not limited to ISO-8859-1.

Comment (rruuaanng) [reply Lib/xml/etree/ElementTree.py]: You are right, this may exist in all 8-bit encodings, such as **ISO-8859-[1,15]**. Please give me some time to test these codes (because it's not easy to find a suitable test text).

Comment (erlend-aasland) [reply Lib/xml/etree/ElementTree.py]: Why only 8-bit encodings? What about UTF-16? Or any other multibyte encoding?

Comment (rruuaanng) [reply Lib/xml/etree/ElementTree.py]: Because this occurs in 8-bit encoding and has been read in non-rb mode (although I can't guarantee that this will not happen in 16-bit, in fact, 16-bit encoding can be read with r instead of rb)

Comment (erlend-aasland) [reply Lib/xml/etree/ElementTree.py]: I suggest closing this PR and going back to the drawing board. Try to understand the issue fully before opening the next PR.

Comment (rruuaanng) [reply Lib/xml/etree/ElementTree.py]: I convert PR into a draft and continue to track the problem.

**Comment by vstinner** [comment]
I closed the PR. The scope of the change is not well defined. Someone has to properly test different encodings and decide which encodings emit a warning or not. The issue should be better analyzed before jumping into a concrete implementation.
